orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py

```diff
@@ -3131,3 +3142,199 @@ def test_debug_options_log_level_validation_fails_on_type_mismatch():
+    inputs = torch.randn(20, 10)
+
+    policy = _fallback._FallbackPolicy.FALLBACK_UNSUPPORTED_DEVICE if matching_policy else _fallback._FallbackPolicy.FALLBACK_UNSUPPORTED_INPUT
+
+    ort_model.train(is_training)
+    pt_model.train(is_training)
+    if fallback_enabled:
+        ort_model._torch_module._execution_manager(is_training=is_training)._fallback_manager._policy = policy
+        if matching_policy:
+            ort_out = ort_model(inputs)
+            pt_out = pt_model(inputs)
+            _test_helpers.assert_values_are_close(ort_out, pt_out, rtol=0, atol=0)
+        else:
+            with pytest.raises(_fallback.ORTModuleFallbackException) as type_error:
+                ort_model(inputs)
+            assert "ORTModule supports a single device per model" in str(type_error.value)
+    else:
+        ort_model._torch_module._execution_manager(is_training=is_training)._fallback_manager._policy = _fallback._FallbackPolicy.FALLBACK_DISABLE
+        with pytest.raises(_fallback.ORTModuleFallbackException) as type_error:
+            ort_model(inputs)
+        assert "ORTModule supports a single device per model" in str(type_error.value)
+
+
+@pytest.mark.parametrize("is_training,fallback_enabled,matching_policy", [(True, True, True), (False, True, True), (True, False, True), (False, False, True),
+                                                                          (True, True, False), (False, True, False), (True, False, False), (False, False, False)])
+def test_ortmodule_fallback_device__mismatch(is_training, fallback_enabled, matching_policy):
+    # is_training: True for torch.nn.Module training model, eval mode otherwise
+    # fallback_enabled: True results in PyTorch executing the forward graph instead of ORT backend
+    # matching_policy: True results in properly matching FALLBACK_UNSUPPORTED_DEVICE policy to ORTModuleDeviceException exception.
+    #   Otherwise, an incorrect policy (FALLBACK_UNSUPPORTED_INPUT) is used to verify that the fallback does not happen
+
+    data_device = 'cuda'
+    N, D_in, H, D_out = 64, 784, 500, 10
+
+    # For initial model export, use same device for data and model
+    model = NeuralNetSinglePositionalArgument(D_in, H, D_out)
+    model = ORTModule(model)
+    model.train(is_training)
+    output = model(torch.randn(N, D_in))
+
+    policy = _fallback._FallbackPolicy.FALLBACK_UNSUPPORTED_DEVICE if matching_policy else _fallback._FallbackPolicy.FALLBACK_UNSUPPORTED_INPUT
+
+    # Use data in different device for testing
+    inputs = torch.randn(N, D_in, device=data_device)
+    if fallback_enabled:
+        model._torch_module._execution_manager(is_training=is_training)._fallback_manager._policy = policy
+        if matching_policy:
+            with pytest.raises(RuntimeError) as e:
+                model(inputs)
+            assert "Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0!" in str(e.value)
```

Comment:
>assert "Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0!" in str(e.value) [](http://example.com/codeflow?start=12&length=119)

so pytorch is also no support such input case?

Reply:
Yes